unmarshal.go

```diff
@@ -398,22 +398,15 @@ func unmarshalValue(rv reflect.Value, v Value, opt *marshalState) error {
 			return participle.Errorf(v.Position(), "expected a map but got %s", v)
 		}
 		t := rv.Type()
-		if t.Key().Kind() != reflect.String {
-			panic(fmt.Sprintf("map keys must be strings but we have %s", t.Key()))
-		}
 		rv.Set(reflect.MakeMap(t))
 		for _, entry := range mapping.Entries {
 			key := reflect.New(t.Key()).Elem()
-			value := reflect.New(t.Elem()).Elem()
-			switch entryKey := entry.Key.(type) {
-			case *String:
-				key.SetString(entryKey.Str)
-			case *Type:
-				key.SetString(entryKey.Type)
-			default:
-				panic(fmt.Errorf("map key must be a string or type but is %s", entry.Key))
+			err := unmarshalValue(key, entry.Key, opt)
```

Comment:
This needs to be stricter than allowing any type, as only hashable types can be keys.

Reply:
Good call! I've updated it to support just strings and numbers now.

Reply:
(Could we even reach this code with a `list`, `map`, or `set` type? I mean: would the _parser_ allow us to construct a map in HCL with any of those as keys?)